firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java

```diff
@@ -48,13 +48,18 @@ public LocalSerializer(RemoteSerializer rpcSerializer) {
   com.google.firebase.firestore.proto.MaybeDocument encodeMaybeDocument(MaybeDocument document) {
     com.google.firebase.firestore.proto.MaybeDocument.Builder builder =
         com.google.firebase.firestore.proto.MaybeDocument.newBuilder();
+
     if (document instanceof NoDocument) {
       NoDocument noDocument = (NoDocument) document;
       builder.setNoDocument(encodeNoDocument(noDocument));
       builder.setHasCommittedMutations(noDocument.hasCommittedMutations());
     } else if (document instanceof Document) {
       Document existingDocument = (Document) document;
-      builder.setDocument(encodeDocument(existingDocument));
+      if (existingDocument.getProto() != null) {
```

Comment:
Optimization 1 ("Avoid encode"): avoid serializing a `Document` to a proto again; instead, keep the serialized version around and reuse it. Storing the proto right in the document object is a strawman.